latch_cli/services/local_dev.py

```diff
@@ -0,0 +1,319 @@
+        pkg_root.joinpath("wf"),
+        pkg_root.joinpath("data"),
+        pkg_root.joinpath("scripts"),
+    ]:
+        if path.exists():
+            coroutines.append(
+                asyncssh.scp(
+                    path,
+                    (ssh_conn, f"~/{pkg_root.name}"),
+                    recurse=True,
+                    block_size=16384,
+                )
+            )
+        else:
+            await aioconsole.aprint(f"Could not find {path} - skipping")
+
+    if len(coroutines) > 0:
+        await asyncio.gather(*coroutines)
+
+
+async def get_messages(ws: websockets.WebSocketClientProtocol, show_output: bool):
+    async for message in ws:
+        msg = json.loads(message)
+        if msg.get("Type") == "exit":
+            return
+        if show_output:
+            await aioconsole.aprint(
+                msg.get("Body"),
+                end="",
+                flush=True,
+            )
+
+
+async def send_message(
+    ws: websockets.WebSocketClientProtocol,
+    message: Union[str, bytes],
+):
+    if isinstance(message, bytes):
+        message = message.decode("utf-8")
+    body = {"Body": message, "Type": None}
+    await ws.send(json.dumps(body))
+    # yield control back to the event loop,
+    # see https://github.com/aaugustin/websockets/issues/867
+    await asyncio.sleep(0)
+
+
+async def run_local_dev_session(pkg_root: Path):
+    scripts_dir = pkg_root.joinpath("scripts").resolve()
+
+    def file_filter(filename: str) -> bool:
+        file_path = Path(filename).resolve()
+        return file_path == scripts_dir or scripts_dir in file_path.parents
+
+    completer = NestedCompleter(
+        {
+            "run-script": PathCompleter(
+                get_paths=lambda: [str(pkg_root)],
+                file_filter=file_filter,
+            ),
+            "run": None,
+            "ls": None,
+            "list-tasks": None,
+            "shell": None,
+        },
+        ignore_case=True,
+    )
+    session = PromptSession(
+        ">>> ",
+        completer=completer,
+        complete_while_typing=True,
+        auto_suggest=AutoSuggestFromHistory(),
+        history=FileHistory(pkg_root.joinpath(".latch_history")),
+    )
+
+    key_path = pkg_root / ".ssh_key"
+
+    with TemporarySSHCredentials(key_path) as ssh:
+        headers = {"Authorization": f"Bearer {retrieve_or_login()}"}
+
+        cent_resp = post(
+            sdk_endpoints["provision-centromere"],
+            headers=headers,
+            json={"public_key": ssh.public_key},
+        )
+
+        init_local_dev_resp = post(
+            sdk_endpoints["local-development"],
+            headers=headers,
+            json={
+                "ws_account_id": current_workspace(),
+                "pkg_root": pkg_root.name,
+            },
+        )
+
+        if init_local_dev_resp.status_code == 403:
+            raise ValueError("You are not authorized to use this feature.")
+
+        cent_data = cent_resp.json()
```

Comment:
would explicitly handle `KeyError` exception like in nucleus when field missing and raise something quickly flagging malformed response body (takes a while to debug otherwise)